### PR TITLE
ARROW-7495: [Java] Remove "empty" concept from ArrowBuf, replace with custom referencemanager

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-7495](https://issues.apache.org/jira/browse/ARROW-7495).

With the introduction of ReferenceManager in the codebase, the need for a separate ArrowBuf is no longer necessary. Instead, once can create a new reference manager that is used for the empty ArrowBuf. For reminder/review, empty arrowbufs have a special behavior in that they don't actually have any reference counting semantics and always stay at one. This allow us to better troubleshoot unallocated memory than what would otherwise be an NPE after calling ValueVector.clear()

### CODE REVIEW

**Comment by tianchen92** [comment]
cc @jacques-n

**Comment by emkornfield** [comment]
Ping @jacques-n do you think you would have time to review this?  @rymurr are you familiar with this part of the code and can do a first pass?

**Comment by rymurr** [comment]
Hey @emkornfield. I'm happy to take a look. I likely won't get to it till Monday as I am traveling at the moment.

**Comment by siddharthteotia** [comment]
Can we please add a test that tells how this will look like? I mean the proposal is to remove the isEmpty flag from ArrowBuf and have the users (whoever wants an empty ArrowBuf) pass a custom reference manager that is as good as an ArrowBuf.

It will be great if we have some test to do this. I am not sure if a new concrete implementation of ReferenceManager should be written to manage empty bufs or can we use the existing ReferenceManager.NO_OP?

**Comment by tianchen92** [comment]
> Can we please add a test that tells how this will look like?
> 
> IIUC, the proposal is to remove the isEmpty flag from ArrowBuf and have the users (whoever wants an empty ArrowBuf) pass a custom reference manager that is as good as an empty ArrowBuf.
> 
> It will be great if we have some test to do this. I am not sure if a new concrete implementation of ReferenceManager should be written to manage empty bufs or can we use the existing ReferenceManager.NO_OP?

Thanks, IMO, one could either use ReferenceManager.NO_OP or write their own ReferenceManager for empty ArrowBuf. I added a simple test case using ReferenceManager.NO_OP.

**Comment by rymurr** [comment]
This looks good to me. Potential nice to have: a static method to generate a default empty buffer of length `n`

**Comment by tianchen92** [comment]
> This looks good to me. Potential nice to have: a static method to generate a default empty buffer of length `n`

Thanks, done.

**Review by rymurr — APPROVED** [review]

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java
Comment (liyafan82) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: I am not sure if directly changing a public API would cause client code break.

Comment (tianchen92) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java:93]: Agree we can't easily change the public api, but i think it's ok if it‘s really needed, and users could correct the behavior when they bump version.

**Comment by siddharthteotia** [comment]
@tianchen92 , can you please revise thia? Looks like the work is completed

**Comment by tianchen92** [comment]
> @tianchen92 , can you please revise thia? Looks like the work is completed

ok, revised, thanks!

**Comment by jacques-n** [comment]
I think it would be good to have a set of tests ensuring that the behavior of the empty buffer is consistent. For example, refcnt, release/allocate, etc. That way, if someone changes the noop reference manager which this is depending on in the future, we don't break the expected behavior of an EmptyBuf as used by all the vector classes. Other than that, this looks good.

**Comment by tianchen92** [comment]
> I think it would be good to have a set of tests ensuring that the behavior of the empty buffer is consistent. For example, refcnt, release/allocate, etc. That way, if someone changes the noop reference manager which this is depending on in the future, we don't break the expected behavior of an EmptyBuf as used by all the vector classes. Other than that, this looks good.

Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.

**Comment by emkornfield** [comment]
@jacques-n are you happy with the test coverage?  Any other issues?

**Comment by jacques-n** [comment]
> Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.

The tests you added are all vector based. I'd like to tests against the "empty" ArrowBuf as well.

**Comment by tianchen92** [comment]
> > Good suggestion, seems empty ArrowBuf was used in vectors before populating data. I have added a set of tests in TestValueVector.
> 
> The tests you added are all vector based. I'd like to tests against the "empty" ArrowBuf as well.

I see, I added more cases in TestArrowBuf

**Comment by jacques-n** [comment]
Looks mostly good. +1

As a follow-on, it'd be nice to include a test which ensures that the Netty Allocator returns an empty-behaving byte buffer when users allocate a zero byte buffer.

**Comment by tianchen92** [comment]
> include a test which ensures that the Netty Allocator returns an empty-behaving byte buffer when users allocate a zero byte buffer

Thanks jacques, I created https://issues.apache.org/jira/browse/ARROW-8870 to track.
@emkornfield Please help merge this PR when you have time, thanks a lot!

**Review by siddharthteotia — APPROVED** [review]

**Comment by siddharthteotia** [comment]
@tianchen92 can you rebase this once? I will merge then.

**Comment by tianchen92** [comment]
> @tianchen92 can you rebase this once? I will merge then.

Thanks, rebased:)

**Pull request merged**
